internal/backend/runtime/omni/controllers/omni/machine.go

```diff
@@ -38,57 +42,116 @@ func NewMachineController() *MachineController {
-					machineRequestSet, err := safe.ReaderGetByID[*omni.MachineRequestSet](ctx, r, machineRequestSetID)
-					if err != nil && !state.IsNotFoundError(err) {
-						return err
-					}
-
-					if machineRequestSet != nil && machineRequestSet.Metadata().Owner() == machineProvisionControllerName {
-						machine.Metadata().Labels().Set(omni.LabelNoManualAllocation, "")
-					}
-				}
-
-				spec := machine.TypedSpec().Value
-
-				spec.ManagementAddress = ipPrefix.Addr().String()
-				spec.Connected = link.TypedSpec().Value.Connected
-
-				machine.Metadata().Labels().Set(omni.MachineAddressLabel, spec.ManagementAddress)
-
-				return nil
-			},
+			TransformFunc: helper.transform,
 		},
 		qtransform.WithExtraMappedInput(
 			qtransform.MapperNone[*omni.MachineRequestSet](),
 		),
+		qtransform.WithExtraMappedInput(
+			qtransform.MapperNone[*infra.ProviderStatus](),
+		),
+		qtransform.WithExtraMappedInput(
+			func(_ context.Context, _ *zap.Logger, _ controller.QRuntime, infraMachine *infra.Machine) ([]resource.Pointer, error) {
+				ptr := siderolink.NewLink(resources.DefaultNamespace, infraMachine.Metadata().ID(), nil).Metadata()
+
+				return []resource.Pointer{ptr}, nil
+			},
+		),
 		qtransform.WithConcurrency(4),
 	)
 }
+
+type machineControllerHelper struct{}
+
+func (h *machineControllerHelper) transform(ctx context.Context, r controller.Reader, _ *zap.Logger, link *siderolink.Link, machine *omni.Machine) error {
+	if err := h.handleInfraProvider(ctx, r, link, machine); err != nil {
+		return err
+	}
+
+	// convert SideroLink subnet to an IP address
+	ipPrefix, err := netip.ParsePrefix(link.TypedSpec().Value.NodeSubnet)
+	if err != nil {
+		return err
+	}
+
+	helpers.CopyLabels(link, machine, omni.LabelMachineRequest, omni.LabelMachineRequestSet)
+
+	spec := machine.TypedSpec().Value
+
+	spec.ManagementAddress = ipPrefix.Addr().String()
+	spec.Connected = link.TypedSpec().Value.Connected
+
+	machine.Metadata().Labels().Set(omni.MachineAddressLabel, spec.ManagementAddress)
+
+	return nil
+}
+
+// handleInfraProvider checks if the link has an infra provider ID annotation, and if so, runs the infra provider handling logic - static or provisioning (regular).
+func (h *machineControllerHelper) handleInfraProvider(ctx context.Context, r controller.Reader, link *siderolink.Link, machine *omni.Machine) error {
+	infraProviderID, ok := link.Metadata().Annotations().Get(omni.LabelInfraProviderID)
+	if !ok { // not created by an infra provider, skip
+		return nil
+	}
+
+	providerStatus, err := safe.ReaderGetByID[*infra.ProviderStatus](ctx, r, infraProviderID)
+	if err != nil {
+		return err
+	}
+
+	if _, isStaticProvider := providerStatus.Metadata().Labels().Get(omni.LabelIsStaticInfraProvider); isStaticProvider { // static infra provider flow, e.g., bare-metal
+		return h.handleStaticInfraProvider(ctx, r, machine)
+	}
+
+	// provisioning (regular) infra provider flow, e.g., kubevirt
+	return h.handleProvisioningInfraProvider(ctx, r, link, machine)
```

Comment:
Other than this branching here, the logic in this file is totally unchanged - just reworked for readability.